core/kernel/tee_ta_manager.c

```diff
@@ -670,6 +673,7 @@ TEE_Result tee_ta_open_session(TEE_ErrorOrigin *err,
 
 	if (!ctx || ctx->panicked) {
 		DMSG("panicked, call tee_ta_close_session()");
+		is_open_ta_session_panicked = true;
```

Comment:
would you mean calling `tee_ta_put_session(s);` here instead?

Reply:
yes,
Thank you for reviewing my patch
Ref_count has a value of 2, when tee_ta_unlink_session is called,

Reply:
Actually I think this change cannot be merged.
`is_open_ta_session_panicked` does not handle threaded execution contexts.

Could you provide a test setup (xtest based like) to evidence the hang?
Maybe we can you use  https://github.com/OP-TEE/optee_test/pull/347 to reproduce your issue. It makes a TA to panic during its open session sequence during `xtest 1021`.

Reply:
xtest 1021: hang

`D/TC:? 0 tee_ta_init_session_with_context:458 Re-open TA a4c04d50-f180-11e8-8eb2-f2801f1b9fd1
D/TC:? 0 tee_ta_open_session:549 panicked, call tee_ta_close_session()
D/TC:? 0 tee_ta_close_session:379 tee_ta_close_session(0x728c41a0)
D/TC:? 0 tee_ta_close_session:398 Destroy session
D/TC:? 0 __wq_rpc:39 sleep thread 0 0x7286d8a0 -2
`

Reply:
When xtest 1021 was executed, it really happened panicked,
The above log is a real case on my machine.
If it is not processed effectively, the system will always sleep,

Reply:
I think I understand the problem now. So the second thread tries to open a session to a TA that already has panicked and will error out here. The problem is really with the interface of the function `tee_ta_close_session()` since it will increase the reference counter via a call to `tee_ta_get_session()`. The best would be to split `tee_ta_close_session()` into two functions. The external `tee_ta_close_session()` will remain as the one called from `syscall_close_ta_session()`, `entry_close_session()` and `free_utc()`.  While in this file we'll use the internal helper function that doesn't call `tee_ta_get_session()` since that's already done. I'll propose something if you don't understand what I mean.

Reply:
split `tee_ta_close_session()` into two functions,  
This is an excellent method.